File: qmux/src/credit.ts
/** Tracks used/max credit for flow control.
 *
 * Mirrors the Rust `Credit` struct. Callers can synchronously try to claim
 * credit, or await until credit becomes available. Calling `close()` causes
 * pending and future `claim()` calls to reject.
 */
export class Credit {
	#used: bigint;
	#max: bigint;
	#closed = false;
	#waiters: Array<{ resolve: () => void; reject: (err: Error) => void }> = [];

	constructor(max: bigint) {
		this.#used = 0n;
		this.#max = max;
	}

	/** Try to claim up to `limit` units. Returns amount claimed (0n if none available). */
	tryClaim(limit: bigint): bigint {
		const available = this.#max - this.#used;
		if (available <= 0n) return 0n;
		const claimed = limit < available ? limit : available;
		this.#used += claimed;
		return claimed;
	}

	/** Claim up to `limit` units, waiting until credit is available.
	 *  Rejects if the credit has been closed. */
	async claim(limit: bigint): Promise<bigint> {
		while (true) {
			if (this.#closed) throw new Error("closed");

			const claimed = this.tryClaim(limit);
			if (claimed > 0n) return claimed;

			await new Promise<void>((resolve, reject) => {
				this.#waiters.push({ resolve, reject });
			});
		}
	}

	/** Return previously claimed credit (for rollback). */
	release(amount: bigint): void {
		this.#used = this.#used > amount ? this.#used - amount : 0n;
		this.#wake();
	}

	/** Increase the max. Returns false if new_max < current max. */
	increaseMax(newMax: bigint): boolean {
		if (newMax < this.#max) return false;
		if (newMax === this.#max) return true;
		this.#max = newMax;
		this.#wake();
		return true;
	}

	/** Close the credit, rejecting all pending and future `claim()` calls. */
	close(): void {
		this.#closed = true;
		const waiters = this.#waiters;
		this.#waiters = [];
		const err = new Error("closed");
		for (const { reject } of waiters) reject(err);
	}

	/** Get current available credit (max - used). */
	get available(): bigint {
		const avail = this.#max - this.#used;
		return avail > 0n ? avail : 0n;
	}

	/** Get the current max value. */
	get max(): bigint {
		return this.#max;
	}

	/** Get the current used value. */
	get used(): bigint {
		return this.#used;
	}

	#wake(): void {
		const waiters = this.#waiters;
		this.#waiters = [];
		for (const { resolve } of waiters) resolve();
	}
}

File: qmux/src/credit.rs
use tokio::sync::watch;

use crate::Error;

#[derive(Debug, Clone, Copy)]
struct CreditState {
    used: u64,
    max: u64,
}

/// Tracks used/max credit for flow control.
///
/// Clone is cheap (watch::Sender is internally Arc'd).
/// Multiple holders can claim/release credit concurrently via `send_if_modified`.
#[derive(Clone, Debug)]
pub struct Credit {
    inner: watch::Sender<CreditState>,
}

impl Credit {
    /// Create with initial max (used starts at 0).
    pub fn new(max: u64) -> Self {
        Self {
            inner: watch::Sender::new(CreditState { used: 0, max }),
        }
    }

    /// Try to claim up to `limit` units. Returns amount claimed (0 if none available).
    pub fn try_claim(&self, limit: u64) -> u64 {
        let mut claimed = 0;
        self.inner.send_if_modified(|state| {
            let available = state.max.saturating_sub(state.used);
            claimed = limit.min(available);
            if claimed > 0 {
                state.used += claimed;
                true
            } else {
                false
            }
        });
        claimed
    }

    /// Claim up to `limit` units, waiting until credit is available.
    /// Returns amount claimed (always > 0 unless the sender is dropped).
    pub async fn claim(&self, limit: u64) -> Result<u64, Error> {
        loop {
            let claimed = self.try_claim(limit);
            if claimed > 0 {
                return Ok(claimed);
            }

            // Wait until state changes (max increases or used decreases)
            let mut rx = self.inner.subscribe();
            rx.wait_for(|state| state.used < state.max)
                .await
                .map_err(|_| Error::Closed)?;
        }
    }

    /// Return previously claimed credit (for rollback).
    pub fn release(&self, amount: u64) {
        self.inner.send_if_modified(|state| {
            state.used = state.used.saturating_sub(amount);
            true
        });
    }

    /// Claim exactly 1 unit and return the index (value of `used` before incrementing).
    /// Waits until credit is available.
    pub async fn claim_index(&self) -> Result<u64, Error> {
        loop {
            let mut index = 0;
            let claimed = self.inner.send_if_modified(|state| {
                if state.used < state.max {
                    index = state.used;
                    state.used += 1;
                    true
                } else {
                    false
                }
            });
            if claimed {
                return Ok(index);
            }

            let mut rx = self.inner.subscribe();
            rx.wait_for(|state| state.used < state.max)
                .await
                .map_err(|_| Error::Closed)?;
        }
    }

    /// Increase the max. Returns error if new_max < current max.
    pub fn increase_max(&self, new_max: u64) -> Result<(), Error> {
        let mut ok = true;
        self.inner.send_if_modified(|state| {
            if new_max < state.max {
                ok = false;
                return false;
            }
            if new_max == state.max {
                return false;
            }
            state.max = new_max;
            true
        });
        if ok {
            Ok(())
        } else {
            Err(Error::FlowControlError)
        }
    }

}
